clients/mesh_rpc_client/src/ws_client.ts

```diff
@@ -0,0 +1,237 @@
+import { assert } from '@0x/assert';
+import { orderParsingUtils } from '@0x/order-utils';
+import { ObjectMap, SignedOrder } from '@0x/types';
+import { BigNumber } from '@0x/utils';
+import * as Web3Providers from '@0x/web3-providers-fork';
+import { v4 as uuid } from 'uuid';
+import * as WebSocket from 'websocket';
+
+import {
+    AcceptedOrderInfo,
+    GetOrdersResponse,
+    HeartbeatEventPayload,
+    OrderEvent,
+    OrderEventPayload,
+    RawAcceptedOrderInfo,
+    RawOrderEvent,
+    RawValidationResults,
+    RejectedOrderInfo,
+    ValidationResults,
+    WSOpts,
+} from './types';
+
+const CLOSE_REASON_NO_HEARTBEAT = 3001;
+const CLOSE_DESCRIPTION_NO_HEARTBEAT = 'No heartbeat received';
+
+const DEFAULT_RECONNECT_AFTER = 5000;
+const DEFAULT_WS_OPTS = {
+    clientConfig: {
+        // For some reason fragmenting the payloads causes the connection to close
+        // Source: https://github.com/theturtle32/WebSocket-Node/issues/359
+        fragmentOutgoingMessages: false,
+    },
+    reconnectAfter: DEFAULT_RECONNECT_AFTER,
+};
+
+/**
+ * This class includes all the functionality related to interacting with a Mesh JSON RPC
+ * websocket endpoint.
+ */
+export class WSClient {
+    private _subscriptionIdToMeshSpecificId: ObjectMap<string>;
+    private _heartbeatCheckIntervalId: number | undefined;
+    private readonly _wsProvider: Web3Providers.WebsocketProvider;
+    private static _convertRawAcceptedOrders(rawAcceptedOrders: RawAcceptedOrderInfo[]): AcceptedOrderInfo[] {
+        const acceptedOrderInfos: AcceptedOrderInfo[] = [];
+        rawAcceptedOrders.forEach(rawAcceptedOrderInfo => {
+            const acceptedOrderInfo: AcceptedOrderInfo = {
+                orderHash: rawAcceptedOrderInfo.orderHash,
+                signedOrder: orderParsingUtils.convertOrderStringFieldsToBigNumber(rawAcceptedOrderInfo.signedOrder),
+                fillableTakerAssetAmount: new BigNumber(rawAcceptedOrderInfo.fillableTakerAssetAmount),
+            };
+            acceptedOrderInfos.push(acceptedOrderInfo);
+        });
+        return acceptedOrderInfos;
+    }
+    /**
+     * Instantiates a new WSClient instance
+     * @param   url               WS server endpoint
+     * @param   wsOpts            WebSocket options
```

Comment:
Maybe we should direct people to where they can find a list of WebSocket options and information about what each option does? Just looking at the doc comment I have no idea what options are available.

Reply:
The method signature has the expected type: `WSOpts`. Going to definition on that type will show you all the options and a description of each.

Reply:
That doesn't work for JavaScript users, right?